File: 1995-finding-pairs-with-a-certain-sum/finding-pairs-with-a-certain-sum.py
class FindSumPairs:

    def __init__(self, nums1: List[int], nums2: List[int]):
        self.nums1 = sorted(nums1)
        self.nums2 = nums2
        self.freq = Counter(nums2)
        self.xMax = max(nums2)

    def add(self, index: int, val: int) -> None:
        old_val = self.nums2[index]
        self.freq[old_val] -= 1
        if self.freq[old_val] == 0:
            del self.freq[old_val]

        self.nums2[index] += val
        new_val = self.nums2[index]
        self.freq[new_val] += 1
        self.xMax = max(self.xMax, new_val)

    def count(self, tot: int) -> int:
        threshold = max(1, tot - self.xMax)
        i0 = bisect_left(self.nums1, threshold)

        count = 0
        for i in range(i0, len(self.nums1)):
            x = self.nums1[i]
            if x >= tot:
                break
            count += self.freq.get(tot - x, 0)
        return count


# Your FindSumPairs object will be instantiated and called as such:
# obj = FindSumPairs(nums1, nums2)
# obj.add(index,val)
# param_2 = obj.count(tot)

File: 1995-finding-pairs-with-a-certain-sum/finding-pairs-with-a-certain-sum.go
type FindSumPairs struct {
    nums1 []int
	nums2 []int
	freq  map[int]int
	xMax  int
}


func Constructor(nums1 []int, nums2 []int) FindSumPairs {
    freq := make(map[int]int)
	xMax := 0

	for _, x := range nums2 {
		freq[x]++
		if x > xMax {
			xMax = x
		}
	}

	sort.Ints(nums1)

	return FindSumPairs{
		nums1: nums1,
		nums2: nums2,
		freq:  freq,
		xMax:  xMax,
	}
}


func (this *FindSumPairs) Add(index int, val int)  {
    oldVal := this.nums2[index]
	this.freq[oldVal]--
	if this.freq[oldVal] == 0 {
		delete(this.freq, oldVal)
	}

	this.nums2[index] += val
	newVal := this.nums2[index]
	this.freq[newVal]++
	if newVal > this.xMax {
		this.xMax = newVal
	}
}


func (this *FindSumPairs) Count(tot int) int {
    count := 0
	threshold := max(1, tot-this.xMax)
	i0 := sort.SearchInts(this.nums1, threshold)

	for i := i0; i < len(this.nums1); i++ {
		y := this.nums1[i]
		if y >= tot {
			break
		}
		count += this.freq[tot-y]
	}
	return count
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}


/**
 * Your FindSumPairs object will be instantiated and called as such:
 * obj := Constructor(nums1, nums2);
 * obj.Add(index,val);
 * param_2 := obj.Count(tot);
 */